pkg/virt-launcher/virtwrap/network/podinterface.go

```diff
@@ -581,7 +581,7 @@ func (b *BridgeBindMechanism) learnInterfaceRoutes() error {
 }
 
 func (b *BridgeBindMechanism) decorateDhcpConfigRoutes(dhcpConfig *cache.DhcpConfig) {
-	dhcpConfig.AdvertisingIPAddr = b.routes[0].Gw
```

Comment:
I wonder how the PR introduced the bug passed e2e tests.

Reply:
I'll dig into this a bit more, but, for the time being, I suspect the cirros based VMs don't care about the missing dhcp-server-identifier option.

Reply:
Yeah, seems like it.

I've added a sleep before [pinging the gateway](), and checked the state of the VMI. It feature an IP address:

```bash
$ _out/cmd/virtctl/virtctl console -nkubevirt-test-default1 testvmi-txhzt        
Successfully connected to testvmi-txhzt console. The escape sequence is ^]

$ ip a s eth0
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast qlen 1000
    link/ether ae:fa:a9:d3:ab:12 brd ff:ff:ff:ff:ff:ff
    inet 10.1.1.7/24 brd 10.1.1.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::acfa:a9ff:fed3:ab12/64 scope link 
       valid_lft forever preferred_lft forever
```

Furthermore, when I remove the IP address and request it again via dhcp, I see it gets a reply:
```bash
$ sudo /sbin/cirros-dhcpc down eth0 
$ sudo /sbin/cirros-dhcpc 
$ ip a s eth0
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast qlen 1000
    link/ether ae:fa:a9:d3:ab:12 brd ff:ff:ff:ff:ff:ff
    inet6 fe80::acfa:a9ff:fed3:ab12/64 scope link 
       valid_lft forever preferred_lft forever
$ sudo /sbin/cirros-dhcpc up eth0 
udhcpc (v1.23.2) started
Sending discover...
Sending select for 10.1.1.7...
Lease of 10.1.1.7 obtained, lease time 86313600
```

This is using kubevirt version:
```
$ _out/cmd/virtctl/virtctl version                                       
Client Version: version.Info{GitVersion:"v0.42.0-rc.0.74+4b634f54a1fd25-dirty", GitCommit:"4b634f54a1fd254ef62c2b96a9bdf043b1682782", GitTreeState:"dirty", BuildDate:"2021-06-07T13:22:26Z", GoVersion:"go1.16.1", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{GitVersion:"v0.42.0-rc.0.74+4b634f54a1fd25", GitCommit:"4b634f54a1fd254ef62c2b96a9bdf043b1682782", GitTreeState:"clean", BuildDate:"2021-06-07T13:09:52Z", GoVersion:"go1.16.1", Compiler:"gc", Platform:"linux/amd64"}
```

I think this shows that cirros DHCP client does not care about the missing `dhcp-server-identifier` option, but if you want, I can back up this claim w/ a traffic capture.

Reply:
OK, it's a different matter. In the captures - running [this test](https://github.com/kubevirt/kubevirt/blob/0adf11b9b41b2fe7aa71f959631553296bbec7d0/tests/network/vmi_multus.go#L174) - I can perfectly see the value for the dhcp-server address, which means that somehow, the pod has a route:

```
15:48:27.222862 IP (tos 0x0, ttl 64, id 0, offset 0, flags [none], proto UDP (17), length 337)
    0.0.0.0.bootpc > 255.255.255.255.bootps: [udp sum ok] BOOTP/DHCP, Request from 5a:5e:3a:42:23:90, length 309, xid 0x147bf619, Flags [none] (0x0000)
	  Client-Ethernet-Address 5a:5e:3a:42:23:90
	  Vendor-rfc1048 Extensions
	    Magic Cookie 0x63825363
	    DHCP-Message Option 53, length 1: Request
	    Client-ID Option 61, length 7: ether 5a:5e:3a:42:23:90
	    Requested-IP Option 50, length 4: 10.1.1.2
	    Server-ID Option 54, length 4: 10.1.1.1
	    MSZ Option 57, length 2: 576
	    Parameter-Request Option 55, length 9: 
	      Subnet-Mask, Default-Gateway, Domain-Name-Server, Hostname
	      Domain-Name, MTU, BR, NTP
	      Classless-Static-Route
	    Vendor-Class Option 60, length 12: "udhcp 1.23.2"
	    Hostname Option 12, length 13: "testvmi-txhzt"
	    END Option 255, length 0
15:48:27.223509 IP (tos 0x0, ttl 64, id 33522, offset 0, flags [DF], proto UDP (17), length 443)
    169.254.75.10.bootps > 255.255.255.255.bootpc: [udp sum ok] BOOTP/DHCP, Reply, length 415, xid 0x147bf619, Flags [none] (0x0000)
	  Your-IP 10.1.1.2
	  Client-Ethernet-Address 5a:5e:3a:42:23:90
	  Vendor-rfc1048 Extensions
	    Magic Cookie 0x63825363
	    DHCP-Message Option 53, length 1: ACK
	    Server-ID Option 54, length 4: 10.1.1.1
	    Lease-Time Option 51, length 4: 86313600
	    Default-Gateway Option 3, length 4: 10.1.1.1
	    Domain-Name-Server Option 6, length 4: 10.96.0.10
	    MTU Option 26, length 2: 1500
	    Classless-Static-Route Option 121, length 0ERROR: length < 5 bytes
	    T119 Option 119, length 76: 376141154,1702259058,1949135973,1936993636,1701208437,1819554051,1937138439,1668052339,1952805381,1819239265,1811940211,1986201443,1819636596,1701971308,1868783980,484204,1970500709,1912958063,1667329024
	    Hostname Option 12, length 13: "testvmi-txhzt"
	    Domain-Name Option 15, length 40: "kubevirt-test-default1.svc.cluster.local"
	    Subnet-Mask Option 1, length 4: 255.255.255.0
	    END Option 255, length 0

```

These are the logs I see in the handler (without this PR's fix, only w/ the added logs):
```
{"component":"virt-handler","level":"info","msg":"got to add 2 routes to the DhcpConfig","pos":"podinterface.go:452","timestamp":"2021-06-07T16:03:39.109823Z"}
{"component":"virt-handler","level":"info","msg":"the default route is: {Ifindex: 6 Dst: 10.1.1.0/24 Src: 10.1.1.4 Gw: 10.1.1.1 Flags: [] Table: 254}","pos":"podinterface.go:587","timestamp":"2021-06-07T16:03:39.109855Z"}
{"component":"virt-handler","level":"info","msg":"route: {Ifindex: 6 Dst: 10.1.1.0/24 Src: 10.1.1.4 Gw: 10.1.1.1 Flags: [] Table: 254}","pos":"common.go:480","timestamp":"2021-06-07T16:03:39.109869Z"}
{"component":"virt-handler","level":"info","msg":"route: {Ifindex: 6 Dst: 10.1.1.1/32 Src: 10.1.1.4 Gw: \u003cnil\u003e Flags: [] Table: 254}","pos":"common.go:480","timestamp":"2021-06-07T16:03:39.109881Z"}
{"component":"virt-handler","level":"info","msg":"The generated dhcpConfig: DhcpConfig: { Name: net1, IPv4: 10.1.1.4/24 net1, IPv6: \u003cnil\u003e, MAC: 3a:d0:3c:a8:f0:a9, AdvertisingIPAddr: 10.1.1.1, MTU: 1500, IPAMDisabled: false}","pos":"podinterface.go:232","timestamp":"2021-06-07T16:03:39.109931Z"}

```

Which makes sense once we check the plugin code (the plugin install these 2 routes): 
https://github.com/containernetworking/plugins/blob/7d8c7676223a68c2620487fefee5127ff9552cb2/plugins/main/ptp/ptp.go#L115

There isn't a test with bridge binding + IPAM; we intend to address this feature in the future - @alonSadan will work on that .

Reply:
I''m confused. The bug was about the `dhcp server identifier` was overridden by the `d.gw`. Isn't it?
We see in the capture that it actually happens the `d.gw` and the `dhcp server identifier` are the same. While the `d.gw` is correct the  `dhcp server identifier` should be the fake address we give to the bridge (prefix - 169.254.75.1).
So I think you first comment was correct and the cirros dhcp client just verifies there is an identifier and don't realy care or use the value.
Please test it (the version prior the fix) with fedora (dhcp client - network manager), I think with network manger dhcp client it would fail.

Reply:
> There isn't a test with bridge binding + IPAM; we intend to address this feature in the future - @alonSadan will work on that 
Is the attached ptp test with IPAM. Otherwise, why do we have dhcp server?

Anyway, we have test for the bridge binding used for the primary network. There it has ipam for sure.

Reply:
> I''m confused. The bug was about the `dhcp server identifier` was overridden by the `d.gw`. Isn't it?

Yes.

> We see in the capture that it actually happens the `d.gw` and the `dhcp server identifier` are the same. While the `d.gw` is correct the `dhcp server identifier` should be the fake address we give to the bridge (prefix - 169.254.75.1).

As I've said above, the capture is executed **without** this patch.

What happened is that we have copied the gateway of the first route we've found, and updated the advertising IP address, and used it both as the dhcp-server-identifier, and as default gateway. That's why the fake bridge address is gone.

> So I think your first comment was correct and the cirros dhcp client just verifies there is an identifier and doesn't really care or use the value.
> Please test it (the version prior the fix) with fedora (dhcp client - network manager), I think that with network manger dhcp client it would fail.

That capture is from before the fix. 

> > There isn't a test with bridge binding + IPAM; we intend to address this feature in the future - @alonSadan will work on that
> 
> Hasn't the attached ptp test IPAM? Otherwise, why do we have dhcp server?

Above, when  I say a test with "bridge binding + IPAM", I meant bridge **CNI** + IPAM.

Do we agree there's a bug when there aren't any routes - or the first route retrieved by netlink doesn't feature a gateway ? 
Whenever that happens, we'll override the advertising IP address with an empty string. 

What I think we're not agreeing, is on the reply of "why did this pass on CI". Our only IPAM tests use bridge binding (in which we copy the gateway of the first route on the pod interface, thus overriding the advertising IP address), with ptp CNI, with host local IPAM.

PTP CNI [configures 2 routes in the pod interface](https://github.com/containernetworking/plugins/blob/7d8c7676223a68c2620487fefee5127ff9552cb2/plugins/main/ptp/ptp.go#L115), one of which features a gateway. Those routes are consistent with the logs I've pasted in the comment above.

What I'm trying to say with this is that the bug I'm trying to fix **only** happens when we either:
  - have no routes in the pod interface
  - the first route in the pod interface does **not** have a gateway defined

PTP does not expose this bug, since it configures a route with a gateway, that for whatever reason happens to be the first in the list (when retrieved).

Reply:
But PTP exposes the bug, the `Server-ID Option` is wrong. Isn't it?

I understand there were two original issue.
1. The route doesn't have `gateway` => the `AdvertisingIPAddr` is overridden with nil.
2. The route has a `gateway` => the `AdvertisingIPAddr` is overridden with `gateway`.

I understand option 1 is not covered by our e2e tests. But how come option 2 which is covered (even in your ptp example) doesn't break the relevant e2e tests?

> have no routes in the pod interface
I believe that in this case we actually don't have a bug since `AdvertisingIPAddr` is not overridden.

Reply:
> But PTP exposes the bug, the `Server-ID Option` is wrong. Isn't it?
> 
> I understand there were two original issues.
> 
> 1. The route doesn't have `gateway` => the `AdvertisingIPAddr` is overridden with `nil`.
> 2. The route has a `gateway` => the `AdvertisingIPAddr` is overridden with `gateway`.
> 
> I understand option 1 is not covered by our e2e tests. But how come option 2 which is covered (even in your ptp example) doesn't break the relevant e2e tests?

I don't have an answer to this, other than the only thing we test is that the cirros VM gets an IP address, and is able to reach the ptp gateway (which it does via the 2 routes it installed via CNI).

> 
> > have no routes in the pod interface
> 
> I believe that in this case we actually don't have a bug since `AdvertisingIPAddr` is not overridden.

+1, you're right in this.